[JENKINS-75905] Include update site URL in signature verification in error messages

adhamahmad opened the pull request:
<!-- Comment:
A great PR typically begins with the line below.
Replace <issue-number> with the issue number.
-->

Fixes #16784

This change improves the readability of signature verification error messages in the plugin manager UI.

- The update site URL is appended to error messages provide context.
- Users now see a concise message in the UI instead of a confusing error box dominated by stack trace details.

<!-- Comment:
If the issue is not fully described in the issue tracker, add more information here (justification, pull request links, etc.).

 * We do not require an issue for minor improvements.
 * Major new features should have an issue created.
-->

### Testing done
- Added a unit test in UpdateSiteTest that simulates a signature verification failure.
- Verified that the resulting FormValidation error message contains the update site URL.
- Manually started Jenkins with a test update site containing an invalid signature as the original issue. Verified in the plugin manager UI that the error message is concise and includes the URL.


<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Screenshots (UI changes only)

<!--
If your change involves a UI change, please include screenshots showing the before and after states.
-->

#### Before
<img width="2540" height="1875" alt="before" src="https://github.com/user-attachments/assets/62851583-170b-4c9b-a176-04540d11ecfb" />

#### After
<img width="1541" height="852" alt="image" src="https://github.com/user-attachments/assets/e6e7ce7d-fbbb-4d04-b8ec-caea47e7d833" />

### Proposed changelog entries

* Display a brief message on update site signature verification failure instead of showing a stack trace.

### Proposed changelog category

/label bug

<!--
The changelog entry needs to have a category which is selected based on the label.
If there's no changelog then the label should be `skip-changelog`.

The available categories are:
* bug - Minor bug. Will be listed after features
* developer - Changes which impact plugin developers
* dependencies - Pull requests that update a dependency
* internal - Internal only change, not user facing
* into-lts - Changes that are backported to the LTS baseline
* localization - Updates localization files
* major-bug - Major bug. Will be highlighted on the top of the changelog
* major-rfe - Major enhancement. Will be highlighted on the top
* rfe - Minor enhancement
* regression-fix - Fixes a regression in one of the previous Jenkins releases
* removed - Removes a feature or a public API
* skip-changelog - Should not be shown in the changelog

Non-changelog categories:
* web-ui - Changes in the web UI

Non-changelog categories require a changelog category but should be used if applicable,
comma separate to provide multiple categories in the label command.
-->

### Proposed upgrade guidelines

N/A

<!-- Comment:
Leave the proposed upgrade guidelines in the pull request with the "N/A" value if no upgrade guidelines are needed.
The changelog generator relies on the presence of the upgrade guidelines section as part of its data extraction process.
-->

### Submitter checklist

- [X] The issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [X] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] UI changes do not introduce regressions when enforcing the current default rules of [Content Security Policy Plugin](https://plugins.jenkins.io/csp/). In particular, new or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

N/A

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- ~If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).~
- ~If it would make sense to backport the change to LTS, be a _Bug_ or _Improvement_, and either the issue or pull request must be labeled as `lts-candidate` to be considered.~

adhamahmad committed:
[JENKINS-75905] Include update site URL in signature verification error messages

adhamahmad committed:
[JENKINS-75905] Include update site URL in signature verification error messages

adhamahmad committed:
[JENKINS-75905] Include update site URL in signature verification error messages

adhamahmad committed:
Merge branch 'closes#16784' of https://github.com/adhamahmad/jenkins into closes#16784

A1exKH reviewed (approved):
@adhamahmad,  LGTM!
Thank you for screenshots and covering it by adding unit test!

MarkEWaite commented on core/src/main/java/jenkins/util/JSONSignatureValidator.java:
```diff
@@ -321,5 +323,25 @@ protected Set<TrustAnchor> loadTrustAnchors(CertificateFactory cf) throws IOExce
+     * @param e the exception to extract the message from
+     * @return a concise, readable error message
+     */
+    private String getRootCauseMessage(Throwable e) {
```
Optional suggestion to annotate the argument so that spotbugs knows that the implementation expects that the argument is not null.  A null pointer exception will be reported if a null value is passed as the argument.
```suggestion
    private String getRootCauseMessage(@NonNull Throwable e) {
```

MarkEWaite reviewed (commented):
Thanks for the pull request.  It looks good in general.

I have a question to clarify the text substitution, since I was unable to find any way to generate an exception message that would require the substitution.

I also offered two optional suggestions.  You are welcome to ignore them if you don't find them to be helpful.

adhamahmad committed:
Merge branch 'closes#16784' of https://github.com/adhamahmad/jenkins into closes#16784

adhamahmad committed:
Use the non-deprecated getJsonSignatureValidator method and add @NonNull annotation

Use the non-deprecated getJsonSignatureValidator Add @NonNull annotation to the Throwable parameter in JSONSignatureValidator#getRootCauseMessage.

adhamahmad commented on core/src/main/java/hudson/model/UpdateSite.java:
```diff
@@ -294,7 +294,29 @@ protected UpdateCenter.InstallationJob createInstallationJob(Plugin plugin, Upda
+                    updatedMessage = message.replaceAll(
+                            "(update site\\s+).*?(\\s+Path)",
+                            "$1" + url + "$2"
+                    );
```
Setting the update site URL to https://jenkins-updates.cloudbees.com/update-center/core-cm/update-center.json as described in the original issue generates a message that contains "update site" and " Path".

MarkEWaite committed:
Test both branches of message formatter

MarkEWaite reviewed (approved):
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback.

/label ready-for-merge
